File: bench/cs/Ssh.Benchmark/SlowStream.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Threading;
using System.Threading.Tasks;

namespace Microsoft.DevTunnels.Ssh.Benchmark;

/// <summary>
/// Simulates additional network latency by inserting async delays in stream writes (not reads).
/// </summary>
class SlowStream : Stream
{
	private Queue<byte[]> writeQueue = new Queue<byte[]>();
	private int writeCount = 0;

	public SlowStream(Stream baseStream, TimeSpan addedLatency)
	{
		BaseStream = baseStream ?? throw new ArgumentNullException(nameof(baseStream));
		AddedLatency = addedLatency;
	}

	public Stream BaseStream { get; }

	public TimeSpan AddedLatency { get; }

	public override bool CanRead => BaseStream.CanRead;

	public override bool CanSeek => BaseStream.CanSeek;

	public override bool CanWrite => BaseStream.CanWrite;

	public override long Length => BaseStream.Length;

	public override long Position { get => BaseStream.Position; set => BaseStream.Position = value; }

	public override int Read(byte[] buffer, int offset, int count) =>
		throw new NotSupportedException("Synchronous stream methods are not supported.");

	public override long Seek(long offset, SeekOrigin origin) => BaseStream.Seek(offset, origin);

	public override void SetLength(long value) => BaseStream.SetLength(value);

	public override void Write(byte[] buffer, int offset, int count) =>
		throw new NotSupportedException("Synchronous stream methods are not supported.");
	public override void Flush() =>
		throw new NotSupportedException("Synchronous stream methods are not supported.");

	public override Task<int> ReadAsync(
		byte[] buffer, int offset, int count, CancellationToken cancellationToken)
	{
		return BaseStream.ReadAsync(buffer, offset, count, cancellationToken);
	}

	public override Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
	{
		var segment = new byte[count];
		Array.Copy(buffer, offset, segment, 0, count);
		lock (this.writeQueue)
		{
			this.writeQueue.Enqueue(segment);
			this.writeCount++;
		}
		return Task.CompletedTask;
	}

	public override Task FlushAsync(CancellationToken cancellationToken)
	{
		int count;
		lock (this.writeQueue)
		{
			// Only flush the number of writes since the last flush.
			// This ensures all writes get exactly the specified delay.
			count = this.writeCount;
			this.writeCount = 0;
		}

		if (count == 0)
		{
			return Task.CompletedTask;
		}

		var stopwatch = new System.Diagnostics.Stopwatch();
		stopwatch.Start();

		Task.Run(async () =>
		{
				// Spinning like this is horribly inefficient, but for benchmarking purposes it enables
				// a much more precise latency simulation compared to something like Task.Delay(ms).
				while (stopwatch.ElapsedMilliseconds < AddedLatency.TotalMilliseconds)
			{
				await Task.Yield();
			}

				// Lock while dequeuing+writing to ensure blocks are written to the base stream in order.
				lock (this.writeQueue)
			{
				while (count-- > 0 && this.writeQueue.TryDequeue(out var buffer))
				{
					BaseStream.Write(buffer, 0, buffer.Length);
				}
			}
		});

		return Task.CompletedTask;
	}
}

File: bench/ts/ssh-bench/slowStream.ts
//
//  Copyright (c) Microsoft Corporation. All rights reserved.
//

import { Event, Emitter, CancellationToken } from 'vscode-jsonrpc';
import { Stream } from '@microsoft/dev-tunnels-ssh';

export class SlowStream implements Stream {
	public constructor(public readonly baseStream: Stream, public readonly addedLatency: number) {}

	private readonly closedEmitter = new Emitter<{ error?: Error }>();
	public readonly closed: Event<{ error?: Error }> = this.closedEmitter.event;

	public async read(count: number, cancellation?: CancellationToken): Promise<Buffer | null> {
		return await this.baseStream.read(count, cancellation);
	}

	public async write(data: Buffer, cancellation?: CancellationToken): Promise<void> {
		const copy = Buffer.from(data);

		function timeMs(): number {
			// Use Node.js high-resolution time API.
			const [s, ns] = process.hrtime();
			return s * 1000 + ns / 1000000;
		}
		const startTime = timeMs();

		new Promise<void>((resolve) => {
			// Spinning like this is horribly inefficient, but for benchmarking purposes it enables
			// a much more precise latency simulation compared to something like setTimeout(f, ms).
			const resolveAfterDelay = () => {
				if (timeMs() - startTime >= this.addedLatency) {
					resolve();
				} else {
					setImmediate(resolveAfterDelay);
				}
			};
			resolveAfterDelay();
		}).then(() => {
			return this.baseStream.write(copy, cancellation);
		});
	}

	public async close(error?: Error, cancellation?: CancellationToken): Promise<void> {
		await this.baseStream.close(error, cancellation);
	}

	public dispose(): void {
		this.baseStream.dispose();
	}

	public get isDisposed(): boolean {
		return this.baseStream.isDisposed;
	}
}
